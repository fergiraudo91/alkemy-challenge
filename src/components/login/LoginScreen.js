import axios from "axios";
import React, { useContext, useState } from "react";
import { AuthContext } from "../../auth/AuthContext";
import { useForm } from "../../hooks/useForm";
import { types } from "../../types/types";
import './login.css';

export const LoginScreen = ({ history }) => {
  const { dispatch } = useContext(AuthContext);
  const lastPath = localStorage.getItem('lastPath') || "/";
  const [error, setError] = useState('');
  const [ values, handleInputChange ] = useForm({email: '', password: ''});
  const callApi = async () => {
    const token = await axios.post('http://challenge-react.alkemy.org/', {
        email: values.email,
        password: values.password
      });
      return token;
  }
  const handleLogin = async (e) => {
    e.preventDefault();
    if (!values.email || !values.password) {
      setError("Please complete the user and password");
    }
    else{
      const token = await callApi()
        .catch(e => {
            if(e.message.includes('401')){
              setError("Your user or password are incorrect");
              return;
            }
          }
        );
      if(!!token){
        localStorage.setItem('token', JSON.stringify(token.data));
        dispatch({
          type: types.login,
          payload: {
            name: values.email
          }
        });
        history.replace(lastPath);
      }
    }

  }
  return (
    <div className="main-login-screen">
      <div className="login-screen">
        <img src="assets/img/batman-ico.png" alt="batman logo" />
        <form className="login-form" onSubmit={handleLogin}>
          <div className="form-group">
            <label htmlFor="exampleInputEmail1" className="field">User</label>
            <input
              type="mail"
              className="form-control"
              id="exampleInputEmail1"
              required
              name="email"
              value={values.email}
              onChange={handleInputChange}
            />
          </div>
          <div className="form-group">
            <label htmlFor="exampleInputPassword1" className="field">Password</label>
            <input
              type="password"
              className="form-control"
              id="exampleInputPassword1"
              required
              name="password"
              value={values.password}
              onChange={handleInputChange}
            />
          </div>
          <button
            type="submit" className="btn btn-light">
            Login
          </button>
        </form>
        {error && <div className="alert alert-danger">{error}</div>}
      </div>
    </div>
  );
};
