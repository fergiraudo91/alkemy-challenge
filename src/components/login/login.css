.login-screen {
  margin: 20px;
  padding: 20px;
  text-align: center;
  background-color: #581845;
  color: orange;
  width: 40%;
  border-radius: 5px;
  box-shadow: 7px 20px 16px -3px rgba(0, 0, 0, 0.75);
  -webkit-box-shadow: 7px 20px 16px -3px rgba(0, 0, 0, 0.75);
  -moz-box-shadow: 7px 20px 16px -3px rgba(0, 0, 0, 0.75);
}

.btn{
    background-color: orange;
    font-weight: bold;
    color: black;
}

.form-control{
    background-color: #A770B2;
}

.login-form {
  width: 100%;
  margin: 0 auto;
  padding: 10px;
}

.main-login-screen {
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 100vh;
  justify-content: center;
  background-color: #E2CEE5;
}

.field{
    display: block;
    text-align: left;
}

@media screen and (max-width: 690px) {
    .login-screen{
        width: 80%;
        margin: 10px;
        padding: 10px;
    }

    
}