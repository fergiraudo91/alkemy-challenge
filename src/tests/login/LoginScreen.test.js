import React from 'react';
import { mount } from "enzyme";
import { LoginScreen } from "../../components/login/LoginScreen";
import { AuthContext } from '../../auth/AuthContext';
import '@testing-library/jest-dom';


describe('Testing <LoginScreen />', () => {
    const dispatch = jest.fn();
    const wrapper = mount(
        <AuthContext.Provider value={{dispatch}}>
            <LoginScreen />
        </AuthContext.Provider>
    );

    test('should show an error if the login data is incorrect', () => {

        const value = 'Hola mundo';
        const userInput = wrapper.find('#exampleInputEmail1');
        const passInput = wrapper.find('#exampleInputPassword1');
        userInput.simulate('change', {target: {value}});
        passInput.simulate('change', {target: {value}});
        const form = wrapper.find('form');
        form.simulate('submit');
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').exists()).toBe(true);   
    });

});
