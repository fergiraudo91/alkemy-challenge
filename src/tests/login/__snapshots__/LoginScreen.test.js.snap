// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <LoginScreen /> should show an error if the login data is incorrect 1`] = `
<div
  className="main-login-screen"
>
  <div
    className="login-screen"
  >
    <img
      alt="batman logo"
      src="assets/img/batman-ico.png"
    />
    <form
      className="login-form"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          className="field"
          htmlFor="exampleInputEmail1"
        >
          User
        </label>
        <input
          className="form-control"
          id="exampleInputEmail1"
          name="email"
          onChange={[Function]}
          required={true}
          type="mail"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          className="field"
          htmlFor="exampleInputPassword1"
        >
          Password
        </label>
        <input
          className="form-control"
          id="exampleInputPassword1"
          name="password"
          onChange={[Function]}
          required={true}
          type="password"
          value=""
        />
      </div>
      <button
        className="btn btn-light"
        type="submit"
      >
        Login
      </button>
    </form>
    <div
      className="alert alert-danger"
    >
      Please complete the user and password
    </div>
  </div>
</div>
`;
